src/python_ml_project_template/utils/script_utils.py

```diff
@@ -60,36 +80,96 @@ def flatten_outputs(outputs: List[TorchTree]) -> TorchTree:
     return cast(TorchTree, output_dict)
 
 
-class LogPredictionSamplesCallback(Callback):
-    def __init__(self, logger: WandbLogger):
-        self.logger = logger
+class CanMakePlots(Protocol):
+    @staticmethod
+    @abc.abstractmethod
+    def make_plots(preds, batch: tgd.Batch):
+        pass
+
+
+class LightningModuleWithPlots(pl.LightningModule, CanMakePlots):
+    pass
+
 
-    def on_validation_batch_end(
-        self, trainer, pl_module, outputs, batch, batch_idx, dataloader_idx=0
+class LogPredictionSamplesCallback(Callback):
+    # def __init__(self, logger: WandbLogger):
+    #     self.logger = logger
+
+    # def on_validation_batch_end(
+    #     self, trainer, pl_module, outputs, batch, batch_idx, dataloader_idx=0
+    # ):
+    #     """Called when the validation batch ends."""
+
+    #     # `outputs` comes from `LightningModule.validation_step`
+    #     # which corresponds to our model predictions in this case
+
+    #     # Let's log 20 sample image predictions from the first batch
+    #     if batch_idx == 0:
+    #         n = 20
+    #         x, y = batch
+    #         images = [img for img in x[:n]]
+    #         outs = outputs["preds"][:n].argmax(dim=1)
+    #         captions = [
+    #             f"Ground Truth: {y_i} - Prediction: {y_pred}"
+    #             for y_i, y_pred in zip(y[:n], outs)
+    #         ]
+
+    #         # Option 1: log images with `WandbLogger.log_image`
+    #         self.logger.log_image(key="sample_images", images=images, caption=captions)
+
+    #         # Option 2: log images and predictions as a W&B Table
+    #         columns = ["image", "ground truth", "prediction"]
+    #         data = [
+    #             [wandb.Image(x_i), y_i, y_pred]
+    #             for x_i, y_i, y_pred in list(zip(x[:n], y[:n], outs))
+    #         ]
+    #         self.logger.log_table(key="sample_table", columns=columns, data=data)
+    def __init__(
+        self, train_dset, val_dset, unseen_dset=None, eval_per_n_epoch: int = 1
+    ):
+        self.train_dset = train_dset
+        self.val_dset = val_dset
+        self.unseen_dset = unseen_dset
+        self.eval_per_n_epoch = eval_per_n_epoch
+
+    @staticmethod
+    def eval_log_random_sample(
+        trainer: pl.Trainer,
+        pl_module: LightningModuleWithPlots,
+        dset,
+        prefix: Literal["train", "val", "unseen"],
     ):
-        """Called when the validation batch ends."""
```

Comment:
Instead of deleting this structure, try and take the outputs of the validation step. i.e. stay closer to the original template code, because holding on to private copies of the datasets is not great, which Ben learned between writing Flowbot and now.